gtk/src/file_filter_info.rs

```diff
@@ -0,0 +1,176 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use std::ffi::CStr;
+
+use glib::translate::*;
+use libc::c_char;
+
+use crate::FileFilterFlags;
+
+#[repr(transparent)]
+#[doc(alias = "GtkFileFilterInfo")]
+pub struct FileFilterInfo(ffi::GtkFileFilterInfo);
+
+impl FileFilterInfo {
+    pub fn new(
+        filename: Option<&str>,
+        uri: Option<&str>,
+        display_name: Option<&str>,
+        mime_type: Option<&str>,
+    ) -> Self {
+        skip_assert_initialized!();
+        let mut contains = FileFilterFlags::empty();
+        for (value, bit) in [
+            (filename, FileFilterFlags::FILENAME),
+            (uri, FileFilterFlags::URI),
+            (display_name, FileFilterFlags::DISPLAY_NAME),
+            (mime_type, FileFilterFlags::MIME_TYPE),
+        ] {
+            if value.is_some() {
+                contains |= bit;
+            }
+        }
+
+        Self(ffi::GtkFileFilterInfo {
+            contains: contains.bits(),
+            filename: filename.to_glib_full(),
+            uri: uri.to_glib_full(),
+            display_name: display_name.to_glib_full(),
+            mime_type: mime_type.to_glib_full(),
+        })
+    }
+
+    #[doc(alias = "get_contains")]
+    pub fn contains(&self) -> FileFilterFlags {
+        FileFilterFlags::from_bits_truncate(self.0.contains)
+    }
+
+    #[doc(alias = "get_filename")]
+    pub fn filename(&self) -> Option<&str> {
+        if self.0.filename.is_null() {
+            None
+        } else {
+            unsafe { CStr::from_ptr(self.0.filename).to_str().ok() }
+        }
+    }
+
+    #[doc(alias = "get_uri")]
+    pub fn uri(&self) -> Option<&str> {
+        if self.0.uri.is_null() {
+            None
+        } else {
+            unsafe { CStr::from_ptr(self.0.uri).to_str().ok() }
+        }
+    }
+
+    #[doc(alias = "get_display_name")]
+    pub fn display_name(&self) -> Option<&str> {
+        if self.0.display_name.is_null() {
+            None
+        } else {
+            unsafe { CStr::from_ptr(self.0.display_name).to_str().ok() }
+        }
+    }
+
+    #[doc(alias = "get_mime_type")]
+    pub fn mime_type(&self) -> Option<&str> {
+        if self.0.mime_type.is_null() {
+            None
+        } else {
+            unsafe { CStr::from_ptr(self.0.mime_type).to_str().ok() }
```

Comment:
The `unsafe` block should also cover the `NULL` check as that `CStr` calls is only OK because you first checked for `NULL` (also elsewhere).

Reply:
Ah, I see.  My general philosophy is to make the `unsafe` block as small as possible, to show exactly what is actually unsafe (checking if a pointer is null or not is not unsafe).  But I'm fine changing it if that's project policy/style.

Reply:
The `CStr::from_ptr()` is only safe because of the `NULL`-check though. We generally try to keep all unsafe code contained in the same block so that outside the block all Rust invariants are kept.